Ds 3206

JustinCCYap opened the pull request.

mwmclean commented on R/stacking.R:
```diff
@@ -0,0 +1,1267 @@
+{
+    result <- lapply(data.set.names, function(nm) {
+        if (!QFileExists(nm))
+            stop("The data file '", nm, "' does not exist in the Display ",
```
`Display` should be `Displayr`

JustinCCYap replied:
Fixed in 9802abe5634b4783532b0460f8c0491eae63646e

oliverbock commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+
+    input.v.names <- input.data.set.metadata$variable.names
+    input.v.labels <- input.data.set.metadata$variable.labels
+    retained.indices <- retainedIndices(stacking.groups,
```
Indices into what space?  I think it's the original data file.  Your variable names need to be more explicit where they are calculated in a remote function and passed around.

JustinCCYap replied:
Improved names in 57a6712

oliverbock commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+    input.v.labels <- input.data.set.metadata$variable.labels
+    retained.indices <- retainedIndices(stacking.groups,
+                                        input.data.set.metadata$n.variables)
+    stacked.indices <- stackedIndices(stacking.groups, retained.indices)
```
Indices into what space?  I think it's the _output_ data file.

JustinCCYap replied:
Improved names in 57a6712

oliverbock commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+
+    stacking.groups <- stackWithCommonLabels(common.labels.list,
+                                             input.data.set.metadata)
+    stacking.groups <- stackManually(stacking.groups, manual.stacking,
+                                     specify.by, input.data.set.metadata)
```
The way you pass down stacking.groups into stackManually(), and then merge the two internally, is confusing.  It would be better to create stacking.groups.from.labels and stacking.groups.manual and then merge the two at the top level.  That simplifies stackManually() and makes it easier to unit test.  It also surfaces a critical part of the algorithm (when they are merged).

JustinCCYap replied:
I have refactored the code as you described and created `mergeCommonLabelAndManualStackingGroups`

oliverbock commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+# Perform stacking given a character vector of common labels.
+# A matrix of variable indices (stacking.group) is returned where each row
+# represents a single stacking of variables into one variable.
+stackWithCommonLabels <- function(common.labels.list, input.data.set.metadata)
```
Unit tests.

JustinCCYap replied:
Added unit tests

oliverbock commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+# Parse user-input manual stacking strings and append the variables to manually
+# stack to stacking.groups. Most of the work is done by either
+# stackingSpecifiedByVariable or stackingSpecifiedByObservation.
+stackManually <- function(stacking.groups, manual.stacking,
```
Unit tests.

JustinCCYap replied:
Added unit tests

oliverbock commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+# Parse user-input manual stacking strings for stacking specified by
+# observation, i.e., each string in manual.stacking describes the variables in
+# an observation to be used in stackings of variables.
+stackingSpecifiedByObservation <- function(stacking.groups, manual.stacking,
```
I don't get the difference between this and stackingSpecifiedByVariable.  Unit tests and examples would have helped.

JustinCCYap replied:
Added unit tests

oliverbock reviewed (commented):
Good structure.  Good variable names, generally.  Good parameter checking.

It took me a long time to get a surface understanding of this code because it is divided over many functions, and R provides for no variable typing.  Please add a comment at the top describing the overall algorithm, and important data types that appear in multiple functions (like stacking.groups, stacked.indices) with simple examples.  Not being very familiar with R also made it hard.

Do you warn when stack groups do not have a consistent size?

As always, my suggestions are just suggestions.  If your deeper knowledge of the code tells you to ignore them, then do.

Please resubmit for review when you have made changes, so I can read the comments I'm asking for, and see some more unit tests.

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+#'   local file or file in the Displayr Cloud Drive
+#' @param stacked.data.set.name Name of the stacked data file to be saved in
+#'   the Displayr Cloud Drive (if run from Displayr) or saved locally.
+#' @param stack.with.common.labels \code{"Automatically"},
+#'   \code{"Using a set of variables to stack as reference"},
+#'   \code{"Using manually input common labels"},
+#'   \code{"Disabled"}.
```
This parameter takes a string which comes directly from the GUI combobox corresponding to this parameter. So it has to be readable since users see this, and it helps us too.

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+#' @title Stack data set
+#' @description Stacks variables in a data set.
+#' @param input.data.set.name Name of data file to stack, either as a path to a
+#'   local file or file in the Displayr Cloud Drive
```
Stacking is not intended to be available in Q. So local files only work when running the code locally. I'll update this line.

JustinCCYap replied:
Revised documentation in be77c61

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+        basename(data.set.name.or.path)
+}
+
+cleanMergedDataSetName <- function(merged.data.set.name, data.set.names)
```
Removed in 0174838, I'll be moving this to the merging code

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+{
+    if (is.null(manual.common.labels))
+    {
+        warning("No common labels were manually supplied. No stacking ",
+                "was conducted using common labels.")
```
I see your point and it makes it hard for me to find warnings/errors too, but I don't like lines that are hundreds of characters wide. Perhaps I'll split it into sentences instead.

JustinCCYap replied:
See a436df8

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+    })
+
+    dup <- which(duplicated(result))
+    for (i in dup)
+        result[i] <- uniqueName(result[i], result[-i], "_")
+
```
I don't think this line takes up too much time and I'm not sure how to easily make it run faster. Anyway probably not worth worrying about.

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+#'   variables to omit. Each string can be a combination of comma-separated
+#'   names, wildcards and ranges. See \code{reference.variables.to.stack} for
+#'   more details on the format.
+#' @param write.data.set Whether to write the stacked data set.
```
I'm not sure if this parameter is needed, I'll remove it to avoid confusion. df34958

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+    v.labels <- input.data.set.metadata$variable.labels
+
+    common.labels.list <- list()
+    for (i in seq_along(reference.variables.to.stack))
```
Done in df34958

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+#' @param manual.stacking If \code{specify.by} is "Variable", this is a
+#'   character vector where each string corresponds to variables to be stacked
+#'   into a new variable. If \code{specify.by} is "Observation", this is a
+#'   character vector where each string corresponds to an observation in the
```
No, see 4b69e5d

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+#'   more details on the format.
+#' @param write.data.set Whether to write the stacked data set.
+#' @param include.stacked.data.set.in.output Whether to include the stacked
+#'   data set in the output.
```
Clarified in 4b69e5d

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+#'   \item \code{omitted.stacked.variables} A character vector of omitted
+#'     stacked variables.
+#'   \item \code{common.labels} A character vector of common labels used for
+#'     stacking.
```
Clarified in 1862012

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+{
+    company.secret <- get0("companySecret")
+    !is.null(company.secret) && company.secret != "UNKNOWN"
+}
```
Moved in 810b3bf

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+                 "cloud drive. Ensure that the data file is in the Displayr ",
+                 "cloud drive and its name has been correctly specified.",
+                 call. = FALSE)
+        QLoadData(nm)
```
I've moved the error message to QLoadData: https://github.com/Displayr/flipAPI/commit/5408df630ce23b89b72bd1c65dbb5758e7050eb8#diff-ca75d73bc702f25a52b0f9e829ea63da56420e53afd7a28155d3367366de9ace

and removed it from flipData: 810b3bf

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+}
+
+# Parses a user-input variable range
+parseRange <- function(range.text, variable.names, purpose, on.fail,
```
Added unit tests

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+# suffixes are removed, the remaining text is either numbers or letters but
+# not both (since enumerations occur with letters or numbers but usually not
+# both).
+matchIndicesBasedOnName <- function(ind.list, nms)
```
Added unit tests

JustinCCYap replied:
Examples in the documentation can be automatically run when checking the package but they are not intended to be used as unit tests

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+# common labels and grouping together variables with common labels and the same
+# prefix and suffix, also considering common prefixes and suffixes in variable
+# names if necessary.
+stackingGroupFromCommonLabels <- function(common.labels, v.labels, v.names)
```
Added unit tests

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+# 3. For each set of candidate common labels, form stacking groups
+# 4. Score stacking groups based on size, removing ones with many missing values
+# 5. Choose common labels with highest score
+automaticCommonLabels <- function(input.data.set.metadata)
```
Added unit tests

JustinCCYap commented on R/stacking.R:
```diff
@@ -0,0 +1,1626 @@
+#'   \item \code{is.saved.to.cloud} Whether the stacked data set was saved to
+#'     the Displayr cloud drive.
+#'  }
+#' @export
```
Added examples in c8ff19f

JustinCCYap commented:
No warning is shown when groups have inconsistent sizes as it is often valid and it should be obvious to the user from the output if this is not what they intended (variables will appear missing in the tables of stacked variables).

I've added more comments and documentation and hopefully addressed all your points.

oliverbock reviewed (approved):
Changes look good.  Thanks
